### PR TITLE
Move contents of a deleted inactive user to admin

### LINKED ISSUE #24432: Address Personal Data Retention for inactive accounts deleting authored page content!
**Is your feature request related to a problem? Please describe.**
I figured out a tricky issue that was happening with a site. The homepage and several other pages were being deleted when wp-cron ran.

I believe WooCommerce was using wp-cron scheduled tasks to cleanup inactive/old users with the hook “woocommerce_cleanup_personal_data”. Personal Data Retention was set to Retain Inactive accounts for 1 year. A former site admin user who was still the “author” of many pages was previously demoted to Subscriber role, and consequently was “cleaned up” by this retention policy. The issue was it seems this process was deleting these users without reassigning the ownership of their content, consequently deleting the site homepage and many other pages.

I realize of course the content ownership should have been reassigned when the user was demoted to avoid this.
(from my post at https://wordpress.org/support/topic/personal-data-retention-for-inactive-accounts-deleted-authored-page-content/#post-11825349 )

**Describe the solution you'd like**
It would be nice in the future if this woocommerce_cleanup_personal_data process could be updated to check if deleted users own any published content prior to deletion, possibly notifying the admin to change content the ownership, and or assign content to another admin user with a recent login before deletion. (or skip automatic deletion of owners who own content)

**Describe alternatives you've considered**
The obvious alternative is for a site admin to disable the user deletion Personal Data Retention.  This is easier in hindsight, and only diagnosed thanks to good backups!

**Additional context**
2019-07-17 12_38_17-WooCommerce settings  — WordPress

### PR DESCRIPTION
**Pull request by glaubersilva**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Move contents of a deleted inactive user to admin

It also sends an email to all site administrators informing the user that
was deleted and the list of contents (posts and pages) previously published
by the deleted user that was assigned to one of the administrators.

Thus, we avoid that these contents are lost when an inactive user (maybe a
old admin or author who had published this before to be demoted to subscriber
or customer role) is automatically deleted by WooCommerce data retention policy.

Closes #24432

### How to test the changes in this Pull Request:

1. Create a new user with the administrator role.
2. Assign (as author) pages and posts to the new administrator user created in the previous step.
3. Change the role of the user created in the first step to subscribe or costumer.
4. Inside the plugin folder, locate and open the file "woocommerce/includes/class-wc-privacy.php" in your code editor. Then comment out the lines from 347 to 361.
5. Type in your browser's address bar "example.com/?privacy=test" (replace "example.com" with the address of your local testing website) after you have already added the following code in a custom plugin or in the " function.php "of your theme:

function wootests_privacy() {
	if ( isset( $_GET['privacy'] ) && 'test' === $_GET['privacy'] && class_exists( 'WC_Privacy' ) && in_array( 'woocommerce/woocommerce.php', apply_filters( 'active_plugins', get_option( 'active_plugins' ) ), true ) ) {		
		$p = new WC_Privacy();
		$p->delete_inactive_accounts();
	}
}
add_action( 'init', 'wootests_privacy' );

6. Check if the user created in step 1 has been deleted and if the content published by him (pages and posts) has been transferred to a site administrator.
7. To test whether the email informing administrators about the change is being sent correctly, use something like "MailHog" in your development environment.
8. After completing the test, remember to uncomment the lines of code commented on in step 4 and remove the custom code that was presented in step 5 of this guide.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Move contents of a deleted inactive user to admin

### CODE REVIEW

File: includes/class-wc-privacy.php
Comment (claudiosanches) [inline includes/class-wc-privacy.php:375]: This is not enough, also we may delete some 3rd party content as well, so probably better to include a filter.

You can include this before the foreach:

```php
$builtin_post_types = get_post_types( array( '_builtin' => true ) );
unset( $builtin_post_types['user_request'] ); // Exclude user_request from data migration.

/**
 * Filter post types that should be assign to an admin user prior user exclusion.
 *
 * @param array $post_types List of post types to assign to an admin. Defaults to WordPress builtin post types.
 */
$post_types_to_assign = apply_filters( 'woocommerce_delete_inactive_account_post_types_to_assign', array_values( $builtin_post_types ) );
```

After you just need to update the query to:
```php
'post_type' => $post_types_to_assign,
```

File: includes/class-wc-privacy.php
Comment (claudiosanches) [inline includes/class-wc-privacy.php:388]: No need to pass all post data if you like to just update one raw.

```php
wp_update_post(
	array(
		'ID'          => $post->ID,
		'post_author' => $admin_users[0]->ID,
	)
);
```

File: includes/class-wc-privacy.php
Comment (claudiosanches) [inline includes/class-wc-privacy.php:398]: Here will require update the "pages and/or posts", maybe change to content and after list all post types migrated, you can get the full list from `$post_types_to_assign`.

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Thanks for your help.
I left some suggestions of things that could get improved.

**Comment by glaubersilva** [comment]
Thank you @claudiosanches ! I just made the changes suggested by you. If you have anything else you can do to improve the code, let me know! :)

**Review by claudiosanches — DISMISSED** [review]
Thanks for your help!

File: includes/class-wc-privacy.php
Comment (rrennick) [inline includes/class-wc-privacy.php:383]: `get_posts` defaults to returning 5 posts - https://developer.wordpress.org/reference/functions/get_posts/. To return all post objects by the user `numberposts` needs to be `-1`.

Comment (glaubersilva) [reply includes/class-wc-privacy.php:383]: Done! =)

File: includes/class-wc-privacy.php
Comment (claudiosanches) [inline includes/class-wc-privacy.php:394]: `wp_delete_user()` already supports to reassign to a new user, so we could save some lines of code.

See: https://developer.wordpress.org/reference/functions/wp_delete_user/

So we could change to:

```
$admin_users = get_users( array( 'role' => 'administrator' ) );
$deleted_user = get_user_by( 'ID', $user_id );
wp_delete_user( $user_id, $admin_users[0]->ID ); // Delete user and reassign content to admin.

$posts_made_prior_list = '<ul>';
foreach ( $posts_made_prior as $post ) {
    $posts_made_prior_list .= '<li><a href="' . esc_attr( $post->guid ) . '">' . esc_html( $post->post_title ) . '</a></li>';
}
$posts_made_prior_list .= '</ul>';
```

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Thanks again for all your help.
There's just one small tweak left to do. After that I think it's ready to merge. Thanks!

File: includes/class-wc-privacy.php
Comment (glaubersilva) [reply includes/class-wc-privacy.php]: Done! :)

File: includes/class-wc-privacy.php
Comment (claudiosanches) [inline includes/class-wc-privacy.php:378]:
```suggestion
			 * @since 5.3.0
			 * @param array $post_types List of post types to assign to an admin. Defaults to WordPress builtin post types.
```

File: includes/class-wc-privacy.php
Comment (claudiosanches) [inline includes/class-wc-privacy.php:372]: Now that we use `wp_delete_user()` to properly reassign posts seems like we don't need it anymore.

We could only query for pages, posts and products, so we don't list all content in the email, but we can place a message saying that more content got replaced.
This will for sure help a little in case the user was a lot of content, since currently we are fetching everything and doesn't sounds very important to list all in this email since you can also check in the back-end later.

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Thanks for all your help.
There's another small change that we could do before merging this PR.

File: includes/class-wc-privacy.php
Comment (glaubersilva) [reply includes/class-wc-privacy.php:372]: Hi @claudiosanches 

Let me check if I understand what you said... 

I should remove it:

```
$builtin_post_types = get_post_types( array( '_builtin' => true ) );
unset( $builtin_post_types['user_request'] ); // Exclude user_request from data migration.


/**
 * Filter post types that should be assign to an admin user prior user exclusion.
 *
 * @param array $post_types List of post types to assign to an admin. Defaults to WordPress builtin post types.
 */
$post_types_to_assign = apply_filters( 'woocommerce_delete_inactive_account_post_types_to_assign', array_values( $builtin_post_types ) );
```

And replace this:

```
$posts_made_prior = get_posts(
	array(
		'post_type'   => $post_types_to_assign,
		'author'      => $user_id,
		'numberposts' => -1,
	)
);
```

By something like this:

```
$posts_made_prior = get_posts(
	array(
		'post_type' => array( 'post', 'page', 'product' ),
		'author'      => $user_id,
		'numberposts' => -1,
	)
);
```

So in the e-mail message, we can say something like _"pages, posts, products, and others.."_.

**1)** Is this? 

**2)** Or we should keep the filter for the post types to assign? I have this doubt, although not make much sense to keep the filter since the **wp_delete_user()** will re-assign all the post types, right?

Awaiting your feedback to proceed! =)

**Comment by Konamiman** [comment]
Assigning a reviewer from the Proton team so that we can unblock this.

**Review by barryhughes — CHANGES_REQUESTED** [review]
Apologies for the rather large delay, here ... this PR unfortunately slipped through the cracks. On review, some further thoughts and considerations:

- Sending out an email could be a nice touch, but I'm not sure we *need* this in order to satisfy #24432 (also, documenting which users were deleted by writing to the log might be a nice thing to do here).
- If we do want to send an email, I think we should:
    - Give thought to doing this via an email class and template, as we do with most other emails.
    - Consider using the designated site admin email address, instead of emailing *all* admin users.
    - Consider sending a single email warning of the reassignment of content (since, we clean up to 20 users in a single sweep, rather than sending out upto 20 emails in close succession). 
    - Consider making it possible to disable the sending of emails, and/or make it possible to alter the email content (via filter hooks, for example).

Simply deleting and telling WordPress to reassign the content, and logging that this took place, could be a simpler starting point, though.

**Comment by samueljseay** [comment]
We are closing this pull request, as the last activity has been quite some time ago and we have not received a response from you. If you are still interested or feel like this is being closed by mistake, please open another PR based on a recent version of the trunk branch.

Thanks!

**Pull request closed without merging**
